File: translations/android-java/app/src/main/java/com/apexkit/sdk/ApexKitRealtimeWSClient.java
package com.apexkit.sdk;

import com.google.gson.Gson;
import okhttp3.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.HashMap;

public class ApexKitRealtimeWSClient {
    private String url;
    private OkHttpClient client;
    private WebSocket webSocket;
    private Gson gson;
    private List<OnEventListener> listeners;

    public interface OnEventListener {
        void onEvent(Map<String, Object> event);
    }

    public ApexKitRealtimeWSClient(String baseUrl, String token) {
        this.url = baseUrl.replace("http", "ws").replaceAll("/$", "") + "/ws";
        this.client = new OkHttpClient();
        this.gson = new Gson();
        this.listeners = new ArrayList<>();
    }

    public void connect() {
        Request request = new Request.Builder().url(url).build();
        webSocket = client.newWebSocket(request, new WebSocketListener() {
            @Override
            public void onMessage(WebSocket webSocket, String text) {
                Map<String, Object> msg = gson.fromJson(text, Map.class);
                for (OnEventListener listener : listeners) {
                    listener.onEvent(msg);
                }
            }
        });
    }

    public void subscribe(Map<String, Object> filter) {
        Map<String, Object> msg = new HashMap<>();
        msg.put("type", "Subscribe");
        msg.put("payload", filter);
        webSocket.send(gson.toJson(msg));
    }

    public void onEvent(OnEventListener listener) {
        listeners.add(listener);
    }

    public void disconnect() {
        if (webSocket != null) {
            webSocket.close(1000, "Goodbye");
        }
    }
}

File: translations/android-kotlin/app/src/main/kotlin/com/apexkit/sdk/ApexKitRealtimeWSClient.kt
package com.apexkit.sdk

import com.google.gson.Gson
import okhttp3.*
import java.util.concurrent.TimeUnit

class ApexKitRealtimeWSClient(baseUrl: String, private val token: String?) {
    private val url = baseUrl.replace("http", "ws").removeSuffix("/") + "/ws"
    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()
    private var webSocket: WebSocket? = null
    private val gson = Gson()
    private val listeners = mutableListOf<(Map<String, Any>) -> Unit>()

    fun connect() {
        val request = Request.Builder().url(url).build()
        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) {
                val msg = gson.fromJson(text, Map::class.java) as Map<String, Any>
                listeners.forEach { it(msg) }
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(1000, null)
            }
        })
    }

    fun subscribe(filter: Map<String, Any>) {
        val msg = mapOf(
            "type" to "Subscribe",
            "payload" to mapOf(
                "collection_id" to filter["collectionId"],
                "record_id" to filter["recordId"],
                "event_type" to filter["eventType"],
                "filter" to filter["dataFilter"],
                "channel" to filter["channel"],
                "custom_event" to filter["customEvent"]
            )
        )
        webSocket?.send(gson.toJson(msg))
    }

    fun sendSignal(channel: String, event: String, data: Any) {
        val msg = mapOf(
            "type" to "Signal",
            "payload" to mapOf("channel" to channel, "event" to event, "data" to data)
        )
        webSocket?.send(gson.toJson(msg))
    }

    fun onEvent(listener: (Map<String, Any>) -> Unit) {
        listeners.add(listener)
    }

    fun disconnect() {
        webSocket?.close(1000, "Goodbye")
    }
}
